tests/30rooms/13anonymousaccess.pl

```diff
@@ -385,6 +385,109 @@ sub check_events
+
+            matrix_set_room_history_visibility( $user, $room_id, "world_readable" );
+         }),
+
+         matrix_create_room( $user,
+            visibility => "public",
+            room_alias_name => "listingtest2",
+         )->then( sub {
+            my ( $room_id ) = @_;
+
+            matrix_set_room_history_visibility( $user, $room_id, "invited" );
+         }),
+
+         matrix_create_room( $user,
+            visibility => "public",
+            room_alias_name => "listingtest3",
+         )->then( sub {
+            my ( $room_id ) = @_;
+
+            matrix_set_room_guest_access( $user, $room_id, "can_join" );
+         }),
+
+         matrix_create_room( $user,
+            visibility => "public",
+            room_alias_name => "listingtest4",
+         )->then( sub {
+            my ( $room_id ) = @_;
+
+            Future->needs_all(
+               matrix_set_room_guest_access( $user, $room_id, "can_join" ),
+               matrix_set_room_history_visibility( $user, $room_id, "world_readable" ),
+            );
+         }),
+      )->then( sub {
+         $http->do_request_json(
+            method => "GET",
+            uri    => "/api/v1/publicRooms",
+         )->then( sub {
+            my ( $body ) = @_;
+
+            log_if_fail "publicRooms", $body;
+
+            require_json_keys( $body, qw( start end chunk ));
+            require_json_list( $body->{chunk} );
+
+            my %seen = (
+               listingtest0 => 0,
+               listingtest1 => 0,
+               listingtest2 => 0,
+               listingtest3 => 0,
+               listingtest4 => 0,
+            );
+
+            foreach my $room ( @{ $body->{chunk} } ) {
+               my $aliases = $room->{aliases};
+               my $world_readable = $room->{world_readable};
+               my $guest_can_join = $room->{guest_can_join};
+
+               foreach my $alias ( @{$aliases} ) {
+                  if( $alias =~ m/^\Q#listingtest0:/ ) {
+                     $seen{listingtest0} = $world_readable == JSON::false && $guest_can_join == JSON::false;
+                  }
+                  elsif( $alias =~ m/^\Q#listingtest1:/ ) {
+                     $seen{listingtest1} = $world_readable == JSON::true && $guest_can_join == JSON::false;
+                  }
+                  elsif( $alias =~ m/^\Q#listingtest2:/ ) {
+                     $seen{listingtest2} = $world_readable == JSON::false && $guest_can_join == JSON::false;
+                  }
+                  elsif( $alias =~ m/^\Q#listingtest3:/ ) {
+                     $seen{listingtest3} = $world_readable == JSON::false && $guest_can_join == JSON::true;
+                  }
+                  elsif( $alias =~ m/^\Q#listingtest4:/ ) {
+                     $seen{listingtest4} = $world_readable == JSON::true && $guest_can_join == JSON::true;
+                  }
+               }
+            }
+
+            while ( my ( $key, $value ) = each( %seen ) ) {
```

Comment:
Don't use `each()`. It's terrible and needs to be taken out and shot.[1]

Instead, this wants to be:

    foreach my $key ( keys %seen ) {
       $seen{$key} or die "Wrong for $key";
    }

---

[1]: To store the iterator position, it abuses a slot on the underlying implementation of the actual hash. While it's technically not too bad on a small locally-scoped lexical here, it makes very brittle patterns where iteration gets broken by seemingly unrelated touches made far away. The `foreach` suggested does not suffer this problem.